src/red_black_tree.jl

```diff
@@ -0,0 +1,327 @@
+# it has unique keys
+# leftChild has keys which are less than the node
+# rightChild has keys which are greater than the node
+# color is true if it's a Red Node, else it's false
+mutable struct RBTreeNode{K}
+    color::Bool 
+    data::Union{K, Nothing}
+    leftChild::Union{Nothing, RBTreeNode{K}}
+    rightChild::Union{Nothing, RBTreeNode{K}}
+    parent::Union{Nothing, RBTreeNode{K}}
+
+    RBTreeNode{K}() where K = new{K}(true, nothing, nothing, nothing, nothing)
+    
+    RBTreeNode{K}(d::K) where K = new{K}(true, d, nothing, nothing, nothing)
+end
+
+RBTreeNode() = RBTreeNode{Any}()
+RBTreeNode(d) = RBTreeNode{Any}(d)
+
+function Base.isequal(x::RBTreeNode{K}, y::RBTreeNode{K})where K
+   return (x.color == y.color && x.data == y.data)
+end
+
+Base.:(==)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = isequal(x, y)
+
+Base.:(!=)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = !isequal(x, y)
```

Comment:
According to docs, unless handling floating point specially one only should do `==`

> 
> ## isequal(x, y)
> 
>   Similar to ==, except for the treatment of floating point numbers and of missing values. isequal
>   treats all floating-point NaN values as equal to each other, treats -0.0 as unequal to 0.0, and
>   missing as equal to missing. Always returns a Bool value.
> 
>   Implementation
>   ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡
> 
>   The default implementation of isequal calls ==, so a type that does not involve floating-point
>   values generally only needs to define ==.
> 
>   isequal is the comparison function used by hash tables (Dict). isequal(x,y) must imply that
>   hash(x) == hash(y).
> 
>   This typically means that types for which a custom == or isequal method exists must implement a
>   corresponding hash method (and vice versa). Collections typically implement isequal by calling
>   isequal recursively on all contents.
> 
>   Scalar types generally do not need to implement isequal separate from ==, unless they represent
>   floating-point numbers amenable to a more efficient implementation than that provided as a
>   generic fallback (based on isnan, signbit, and ==).
> 
> ## !=(x, y)
>   ≠(x,y)
> 
>   Not-equals comparison operator. Always gives the opposite answer as ==.
> 
>   Implementation
>   ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡
> 
>   New types should generally not implement this, and rely on the fallback definition !=(x,y) =
>   !(x==y) instead.


So one should just implement `==` here.

and one should implement `hash` if ever implementing `==` (or `isequals`).
I suggest an approproate hash is to just return the hash of the `data`.
The colour won't provide enough gain i think. Worst case is if something with each color is put into the same hashtable, and then is still resolved with a single comparison.